### DISCUSSION TITLE
MENU_TIMEOUT=-1 does not seem a valid option

### DISCUSSION THREAD

**Discussion by intelliant01** [ltsp/ltsp#522] [Q&A] [answered]
@alkisg ```MENU_TIMEOUT=-1``` as documented [herein](https://github.com/ltsp/ltsp/blob/f85c2544e439bf84adc56de84f2bb813a97ae1ee/ltsp/server/ipxe/ltsp.ipxe#L21) produces error in ipxe boot process because ```choose ... --timeout ....``` doesn't take a negative value as documented here - https://ipxe.org/cmd/choose 
[This line](https://github.com/ltsp/ltsp/blob/f85c2544e439bf84adc56de84f2bb813a97ae1ee/ltsp/server/ipxe/ltsp.ipxe#L29) `iseq "${menu-timeout}" "-1" && goto ${img} ||` in `ltsp.ipxe` doesn't seem to be skipping the menu display.

**Comment by alkisg** [comment]
The "-1" value is indeed not supported by ipxe. It's supported by ltsp.ipxe:

https://github.com/ltsp/ltsp/blob/f85c2544e439bf84adc56de84f2bb813a97ae1ee/ltsp/server/ipxe/ltsp.ipxe#L29

So at that point, a comparison with -1 is made, and the control is transferred below if it's equal.
If for some reason this isn't working as expected, we could fix it. I'll check it tomorrow.

**Comment by alkisg** [comment]
It works for me.
Did you put it under [server] or under [common] in ltsp.conf?
Did you run `ltsp ipxe`?
Does /srv/tftp/ltsp/ltsp.ipxe have `isset ${menu-timeout} || set menu-timeout -1` now?

**Reply by intelliant01** [reply]
>Did you put it under [server] or under [common] in ltsp.conf? 

[server]

>Did you run ltsp ipxe?

Yes

>Does /srv/tftp/ltsp/ltsp.ipxe have `isset ${menu-timeout} || set menu-timeout -1 now?`

Yes.

I shall try it again in an hour.

**Reply by intelliant01** [reply]
@alkisg I run into this -
![WhatsApp Image 2021-09-07 at 14 10 54](https://user-images.githubusercontent.com/2577060/132314996-ae7445c2-bed9-4890-be05-51c47ecfe980.jpeg)

Herein are my ltsp.conf and ltsp.ipxe.

[ltsp.conf](https://github.com/ltsp/ltsp/files/7120224/ltsp.conf.txt)
[ltsp.ipxe](https://github.com/ltsp/ltsp/files/7120225/ltsp.ipxe.txt)

**Comment by cyberorg** [comment]
The issue is here: cpuid --ext 29 && set img x86_64 || set img x86_32

Renaming the xyz.img to x86_64.img works.

The script should look if those images exist, then look in ltsp.conf if any default is set, if not then the first one in /srv/ltsp/images/ can be used.

**Reply by intelliant01** [reply]
@cyberorg I tried a `ln -s amd64_ipxe.img x86_64.img` but it did not resolve this. 
Neither did `mv amd64_ipxe.img x86_64.img`.
My `BASE_DIR="/opt/ltsp"`.

**Comment by alkisg** [comment]
@cyberorg, I don't think that line is at all related to MENU_TIMEOUT.

That line is used to automatically choose a 64bit or 32bit image.
Then if these names do not exist, it'll fall back to the values in the menu anyway.

**Reply by cyberorg** [reply]
Fallback doesn't work, that is the reason for what @intelliant01 sees.

This also will work with -1. Set xyz to image that exists.

set img xyz

**Comment by alkisg** [comment]
@intelliant01: open an ipxe shell and test the following commands:

```shell
isset ${menu-timeout} || set menu-timeout -1
iseq "${menu-timeout}" "-1" && echo WORKS ||
```

If these do not work for you, try to update your ipxe version.

**Reply by intelliant01** [reply]
Works.

![WhatsApp Image 2021-09-07 at 15 41 34](https://user-images.githubusercontent.com/2577060/132327999-92903493-8920-4b1c-a46d-0df44102e0d6.jpeg)

So no need to update ipxe.

**Comment by intelliant01** [comment]
As @cyberorg rightfully pointed out, fallback target "x86_64" i.e. the implicit value of `DEFAULT_IMAGE` was missing. 

So if ltsp.conf has `DEFAULT_IMAGE="amd64_ipxe"` along with `MENU_TIMEOUT="-1"` defined in [common] section, then all works fine. 

Thus some method is needed, when a user defines `MENU_TIMEOUT="-1"` in `ltsp.conf` and executes `ltsp ipxe`, to verify and enforce that the value set (implicitly or explcitly) for `DEFAULT_IMAGE` in fact exists as an image.

**Comment by alkisg** [comment, marked as answer]
(you just posted but I already wrote the message, so I'm sending it anyway):

Ah I think I got it. So, you're hiding the menu by using MENU_TIMEOUT=-1, but you didn't set a default image using DEFAULT_IMAGE, **and** your image isn't named x86_64. Then yeah that would fail as $img won't get the correct image name.

If so, the correct solution would be to put this in ltsp.conf:

```shell
[common]
DEFAULT_IMAGE=your-image-name
```

...and then to run `ltsp ipxe`.

**Reply by intelliant01** [reply]
Verification for this needs to be handled in `ltsp ipxe` logic.
